File: solved/2483.MinimumPenaltyForAShop.py
class Solution(object):
  def bestClosingTime(self, customers):
    """
    :type customers: str
    :rtype: int
    """

    N = len(customers)

    clients = []
    empty = []

    ccount = 0
    ecount = 0

    j = N-1
    for i in range(N):

      empty.append(ecount)
      ccount += 1 if customers[j] == 'Y' else 0
      ecount += 1 if customers[i] == 'N' else 0
      clients.append(ccount)

      j -= 1

    clients.reverse()

    min = empty[0] + clients[0]
    minI = 0
    for i in range(N):
      pen = empty[i] + clients[i]
      if pen < min:
        min = pen
        minI = i

    # edge case best time greather than array length
    if calculate(customers, N) < min:
      minI = N
    
    print(minI)
    return minI


def calculate(customers, closing):
  penalty = 0

  for i in range(len(customers)):
    if (i < closing):
      penalty += 1 if customers[i] == 'N' else 0
    else:
      penalty += 1 if customers[i] == 'Y' else 0

  return penalty

Solution.bestClosingTime(None, "YNYY"); # 4
Solution.bestClosingTime(None, "YYNY"); # 2
Solution.bestClosingTime(None, "NYYYYNYN"); # 5
Solution.bestClosingTime(None, "YYYY"); # 4
Solution.bestClosingTime(None, "NNNNN"); # 0

File: solved/2483.MinimumPenaltyForAShop.cpp
#include <string>
#include <iostream>
using namespace std;

class Solution {
public:

  int bestClosingTime(string customers) {
    // cout << '\n' << customers << '\n';

    uint N = customers.length();

    uint clients[N];
    uint empty[N];

    uint ccount = 0;
    uint ecount = 0;
    
    for (uint i = 0, j = N-1; i < N; i++, j--) {
      empty[i] = ecount;
      ccount += customers[j] == 'Y' ? 1 : 0;
      ecount += customers[i] == 'N' ? 1 : 0;
      clients[j] = ccount;
    }

    uint min = empty[0] + clients[0];
    uint minI = 0;
    for (uint i = 0; i < N; i++) {
      uint pen = empty[i] + clients[i];
      if (pen < min) {
        min = pen;
        minI = i;
      }
    }

    // edge case best time greather than array length
    if (calculatePenalty(customers, N) < min) {
      minI = N;
    }

    cout << minI << '\n';

    return minI;
  }

  int calculatePenalty(string customers, int closing) {
    int penalty = 0;

    for (uint i = 0; i < customers.length(); i++) {
      if (i < closing) {
        penalty += customers[i] == 'N' ? 1 : 0;
      }
      else {
        penalty += customers[i] == 'Y' ? 1 : 0;
      }
    }

    return penalty;
  }

};

int main() {
  Solution* s = new Solution();

  s->bestClosingTime("YNYY"); // 4
  s->bestClosingTime("YYNY"); // 2
  s->bestClosingTime("NYYYYNYN"); // 5
  s->bestClosingTime("YYYY"); // 4
  s->bestClosingTime("NNNNN"); // 0

}
